tlsn/tlsn-core/src/serialize.rs

```diff
@@ -0,0 +1,178 @@
+//! Canonical serialization of TLSNotary types.
+
+use crate::{
+    attestation::{AttestationHeader, Field, ATTESTATION_ID_LEN, ATTESTATION_VERSION_LEN},
+    conn::{
+        ConnectionInfo, HandshakeData, HandshakeDataV1_2, ServerEphemKey, ServerSignature,
+        TlsVersion,
+    },
+    encoding::EncodingLeaf,
+    hash::PlaintextHash,
+    transcript::{SliceIdx, SubsequenceIdx},
+};
+
+/// Canonical serialization of TLSNotary types.
+///
+/// This trait is used to serialize types into a canonical byte representation.
+///
+/// It is critical that the serialization is deterministic and unambiguous.
```

Comment:
How do we ensure this? Is this ensured by carefully implementing the trait?  Just asking because we could enforce this through the type system, like using some length prefix encoding for the fields.